### PR TITLE
Extract engine bundle, sign, reattach; also do vsts agent cleanup

### PR DESCRIPTION
**Pull request by chcosta**
Fixes https://github.com/dotnet/core-setup/issues/2195

This is a minimal change, it just creates targets we directly call in the build which wrap calling `insignia.exe` to extract the engine from the bundle, then we sign it, then another wrapper call to reattach the signed engine to the bundle.

The diff for the build definition looks larger because I'm deleting some steps that were disabled.

### CODE REVIEW

File: buildpipeline/Core-Setup-Signing-Windows-BT.json [hunk 1]
```diff
@@ -535,41 +521,49 @@
...
         "versionSpec": "1.*",
         "definitionType": "task"
       },
       "inputs": {
-        "CopyRoot": "",
-        "Contents": "**\\*.log",
-        "ArtifactName": "Build Logs",
-        "ArtifactType": "Container",
-        "TargetPath": "\\\\my\\share\\$(Build.DefinitionName)\\$(Build.BuildNumber)"
+        "solution": "$(PB_SourcesDirectory)\\Tools\\scripts\\vstsagent\\cleanupagent.proj",
```
Comment (eerhardt) [inline buildpipeline/Core-Setup-Signing-Windows-BT.json]: Why is this necessary?

File: buildpipeline/Core-Setup-Signing-Windows-BT.json [hunk 2]
```diff
@@ -411,85 +373,109 @@
...
         "versionSpec": "1.*",
         "definitionType": "task"
       },
       "inputs": {
-        "solution": "build_projects\\signing\\sign.proj",
-        "platform": "",
+        "solution": "$(PB_SourcesDirectory)\\src\\pkg\\packaging\\dir.proj",
+        "platform": "$(PB_TargetArchitecture)",
         "configuration": "$(BuildConfiguration)",
-        "msbuildArguments": "/t:SignBundle $(MsbuildSigningArguments)",
```
Comment (eerhardt) [inline buildpipeline/Core-Setup-Signing-Windows-BT.json]: If we don't sign the bundle here, where does it get signed?

Comment (eerhardt) [inline buildpipeline/Core-Setup-Signing-Windows-BT.json]: In the old world the steps went:

Compile
Sign Assemblies
Test and Package (would create .msi)
Sign MSI and cab
Generate Bundle
Extract Engine from bundle
Sign Engine
Reattach Engine to Bundle
Sign Bundle
Publish


Do your new steps happen in the same order?

---
In reply to: [114195530](https://github.com/dotnet/core-setup/pull/2235#discussion_r114195530) [](ancestors = 114195530)

File: buildpipeline/Core-Setup-Signing-Windows-BT.json (hunk 1 above)
Comment (chcosta) [inline buildpipeline/Core-Setup-Signing-Windows-BT.json]: This is not specifically related to the engine extract / sign / reattach stuff.  But, we've found that VSTS tends to leave builds around and after ~6 weeks, machines will hit out of disk space issues.  So we have this tooling around cleaning up build agents.

Comment (eerhardt) [inline buildpipeline/Core-Setup-Signing-Windows-BT.json]: Just for windows?

File: buildpipeline/Core-Setup-Signing-Windows-BT.json (hunk 2 above)
Comment (chcosta) [inline buildpipeline/Core-Setup-Signing-Windows-BT.json]: This diff is a bit confusing.  The build definition was cloned directly from the old definition, and then steps were disabled and replaced with the new steps in the new system.  This delete is just cleaning up the disabled steps from the definition `"enabled": "false"`.  The steps should be similar to the previous system:

Compile
Sign HostFxr
Build Nuget packages
Build managed / publish shared framework
Sign crossgened assemblies
sign assemblies *.dll
package
build and run tests
sign msi and cab
generate bundle
extract engine
sign engine
reattach engine
sign bundle
publish

Comment (eerhardt) [inline buildpipeline/Core-Setup-Signing-Windows-BT.json]: That's not the order I see when looking at the file:  https://github.com/chcosta/core-setup/blob/6c6978960e732857364f7475aea8759a8f8e2432/buildpipeline/Core-Setup-Signing-Windows-BT.json

I see:

Build nuget packages
Build managed / publish sharedframework
Sign Shared Framework Crossgened Assemblies
Sign Assemblies: *.dll
Package
Build and run tests
Sign MSI and cab
**Extract engine from bundle**
**Sign Bundle**
**Reattach engine to bundle**
**Sign engine**
Publish


Notice:  

- No generate bundle
- And signing the bundle before reattaching the engine to it

Comment (chcosta) [inline buildpipeline/Core-Setup-Signing-Windows-BT.json]: When I load that file into VSTS, here are the build steps, in order...

image

Generate bundle isn't an explicit step anymore because I noticed it was already happening as part of building installers - https://github.com/dotnet/core-setup/blob/master/src/pkg/packaging/dir.proj#L147

Comment (chcosta) [inline buildpipeline/Core-Setup-Signing-Windows-BT.json]: If the order is wrong, I'm completely open to fixing it, I just want to make sure we're looking at the same thing.

Comment (eerhardt) [inline buildpipeline/Core-Setup-Signing-Windows-BT.json]: I'm just looking at the file as it exists in the PR.  I don't know how the ordering happens...  I assumed the order of the file was the order they ran.

Comment (chcosta) [inline buildpipeline/Core-Setup-Signing-Windows-BT.json]: The order in the file appears to be the same as what I am seeing in the image above.

Comment (eerhardt) [inline buildpipeline/Core-Setup-Signing-Windows-BT.json]: Your image above has the wrong ordering as well.

image

You can't sign the bundle before you reattach the engine to it.  That makes the signing invalid.

Please see my comment above for the correct order of things.

Comment (chcosta) [inline buildpipeline/Core-Setup-Signing-Windows-BT.json]: ah, now I see what you're referring to.  Yes, that makes perfect sense.

File: buildpipeline/Core-Setup-Signing-Windows-BT.json (hunk 1 above)
Comment (chcosta) [inline buildpipeline/Core-Setup-Signing-Windows-BT.json]: For all platforms.  I had previously already added this step to Linux and osx.

File: src/pkg/packaging/dir.props [hunk 3]
```diff
@@ -25,6 +25,7 @@
 
   <PropertyGroup>
     <CombinedInstallerFile>dotnet-$(ProductMoniker)$(CombinedInstallerExtension)</CombinedInstallerFile>
+    <CombinedInstallerEngine>dotnet-$(ProductMoniker)-engine.exe</CombinedInstallerEngine>
```
Comment (eerhardt) [inline src/pkg/packaging/dir.props]: I was looking before, and I don't see the below property `SharedFrameworkEngine` being used.  Can it be removed?

File: buildpipeline/Core-Setup-Signing-Windows-BT.json [hunk 4]
```diff
@@ -489,7 +475,7 @@
         "logProjectEvents": "false",
         "createLogFile": "false",
         "msbuildLocationMethod": "version",
-        "msbuildVersion": "15.0",
+        "msbuildVersion": "latest",
```
Comment (eerhardt) [inline buildpipeline/Core-Setup-Signing-Windows-BT.json]: How come this one is `latest` but others are `15.0`?

File: buildpipeline/Core-Setup-Signing-Windows-BT.json [hunk 5]
```diff
@@ -411,64 +373,88 @@
...
+        "configuration": "$(BuildConfiguration)",
+        "msbuildArguments": "/t:ReattachEngineToBundle $(PB_CommonMSBuildArgs) /flp:v=detailed;LogFile=$(PB_SourcesDirectory)\\reattachengine.log",
+        "clean": "false",
+        "maximumCpuCount": "false",
+        "restoreNugetPackages": "false",
+        "logProjectEvents": "false",
+        "createLogFile": "false",
+        "msbuildLocationMethod": "version",
+        "msbuildVersion": "15.0",
+        "msbuildArchitecture": "x86",
```
Comment (eerhardt) [inline buildpipeline/Core-Setup-Signing-Windows-BT.json]: Why are some `x64` and others `x86`?

File: buildpipeline/Core-Setup-Signing-Windows-BT.json (hunk 4 above)
Comment (chcosta) [inline buildpipeline/Core-Setup-Signing-Windows-BT.json]: I've been moving them all to latest in this and other pr's.  specifying 15.0 causes a warning and then it just defaults to "latest".  This caused noise that bothered a bunch of people that look at results on mission control.

File: buildpipeline/Core-Setup-Signing-Windows-BT.json (hunk 5 above)
Comment (chcosta) [inline buildpipeline/Core-Setup-Signing-Windows-BT.json]: ditto as above, I've been moving them all to x64.  I'll load the build definition again and double check the values for all of the steps.
